Recursive update during beans context initalization gets back

In `2.0.0.RC2`.

The problem is widely described in #2140 and #2099. It was fixed in `1.2.3` but it seems it's now present again in `2.0.0.RC2`.

This is **not** a problem of having circular dependencies between beans, because it is not 100% reproducible, maybe 10% tops. This is some race condition during beans context initialization, which was previously fixed successfully for some time.

My example exception is [contained in this comment](https://github.com/micronaut-projects/micronaut-core/issues/2140#issuecomment-650163797).

I'm getting a similar error with Micronaut `2.0.0`:

```
io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type  [io.micronaut.runtime.context.CompositeMessageSource]

Message: Recursive update
Path Taken: new DefaultValidator([ValidatorConfiguration configuration]) --> DefaultValidatorConfiguration.setMessageSource([MessageSource messageSource]) --> new CompositeMessageSource([Collection messageSources])
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1915)
	at io.micronaut.context.DefaultBeanContext.getScopedBeanForDefinition(DefaultBeanContext.java:2386)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2301)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2273)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1233)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForMethodArgument(AbstractBeanDefinition.java:1656)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForMethodArgument(AbstractBeanDefinition.java:899)
	at io.micronaut.validation.validator.$DefaultValidatorConfigurationDefinition.injectBean(Unknown Source)
	at io.micronaut.validation.validator.$DefaultValidatorConfigurationDefinition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1888)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2630)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2616)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2299)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2273)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1233)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1035)
	at io.micronaut.validation.validator.$DefaultValidatorDefinition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1888)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2630)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2616)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2299)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2273)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:1254)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:738)
	at io.micronaut.context.BeanLocator.findBean(BeanLocator.java:149)
	at io.micronaut.context.DefaultBeanContext.getBeanValidator(DefaultBeanContext.java:639)
	at io.micronaut.inject.ValidatedBeanDefinition.validate(ValidatedBeanDefinition.java:43)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1955)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2630)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2616)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2299)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2273)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1233)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1035)
	at io.micronaut.configuration.hibernate.jpa.$JpaConfigurationDefinition.build(Unknown Source)
	at io.micronaut.context.BeanDefinitionDelegate.build(BeanDefinitionDelegate.java:143)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1888)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2630)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2616)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2299)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2273)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:1254)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:738)
	at io.micronaut.configuration.hibernate.jpa.condition.EntitiesInPackageCondition.matches(EntitiesInPackageCondition.java:54)
	at io.micronaut.context.RequiresCondition.matchesCustomConditions(RequiresCondition.java:356)
	at io.micronaut.context.RequiresCondition.processPostStartRequirements(RequiresCondition.java:219)
	at io.micronaut.context.RequiresCondition.matches(RequiresCondition.java:107)
	at io.micronaut.context.AbstractBeanContextConditional.isEnabled(AbstractBeanContextConditional.java:50)
	at io.micronaut.context.AbstractBeanDefinition.isEnabled(AbstractBeanDefinition.java:86)
	at io.micronaut.context.DefaultBeanContext.lambda$findBeanCandidates$37(DefaultBeanContext.java:1701)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.concurrent.ConcurrentLinkedQueue.forEachFrom(ConcurrentLinkedQueue.java:1037)
	at java.base/java.util.concurrent.ConcurrentLinkedQueue$CLQSpliterator.forEachRemaining(ConcurrentLinkedQueue.java:894)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.micronaut.context.DefaultBeanContext.findBeanCandidates(DefaultBeanContext.java:1702)
	at io.micronaut.context.DefaultApplicationContext.findBeanCandidates(DefaultApplicationContext.java:225)
	at io.micronaut.context.DefaultBeanContext.isCandidatePresent(DefaultBeanContext.java:2972)
	at io.micronaut.context.DefaultBeanContext.containsBean(DefaultBeanContext.java:705)
	at io.micronaut.context.BeanDefinitionRegistry.containsBean(BeanDefinitionRegistry.java:355)
	at io.micronaut.context.RequiresCondition.matchesPresenceOfBeans(RequiresCondition.java:543)
	at io.micronaut.context.RequiresCondition.processPostStartRequirements(RequiresCondition.java:211)
	at io.micronaut.context.RequiresCondition.matches(RequiresCondition.java:107)
	at io.micronaut.context.AbstractBeanContextConditional.isEnabled(AbstractBeanContextConditional.java:50)
	at io.micronaut.context.AbstractBeanDefinition.isEnabled(AbstractBeanDefinition.java:86)
	at io.micronaut.context.DefaultBeanContext.lambda$findBeanCandidates$37(DefaultBeanContext.java:1701)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.concurrent.ConcurrentLinkedQueue.forEachFrom(ConcurrentLinkedQueue.java:1037)
	at java.base/java.util.concurrent.ConcurrentLinkedQueue$CLQSpliterator.forEachRemaining(ConcurrentLinkedQueue.java:894)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.micronaut.context.DefaultBeanContext.findBeanCandidates(DefaultBeanContext.java:1702)
	at io.micronaut.context.DefaultApplicationContext.findBeanCandidates(DefaultApplicationContext.java:225)
	at io.micronaut.context.DefaultBeanContext.isCandidatePresent(DefaultBeanContext.java:2972)
	at io.micronaut.context.DefaultBeanContext.containsBean(DefaultBeanContext.java:705)
	at io.micronaut.context.BeanDefinitionRegistry.containsBean(BeanDefinitionRegistry.java:355)
	at io.micronaut.context.RequiresCondition.matchesPresenceOfBeans(RequiresCondition.java:543)
	at io.micronaut.context.RequiresCondition.processPostStartRequirements(RequiresCondition.java:211)
	at io.micronaut.context.RequiresCondition.matches(RequiresCondition.java:107)
	at io.micronaut.context.AbstractBeanContextConditional.isEnabled(AbstractBeanContextConditional.java:50)
	at io.micronaut.context.AbstractBeanDefinition.isEnabled(AbstractBeanDefinition.java:86)
	at io.micronaut.context.DefaultBeanContext.lambda$findBeanCandidates$37(DefaultBeanContext.java:1701)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.concurrent.ConcurrentLinkedQueue.forEachFrom(ConcurrentLinkedQueue.java:1037)
	at java.base/java.util.concurrent.ConcurrentLinkedQueue$CLQSpliterator.forEachRemaining(ConcurrentLinkedQueue.java:894)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.micronaut.context.DefaultBeanContext.findBeanCandidates(DefaultBeanContext.java:1702)
	at io.micronaut.context.DefaultApplicationContext.findBeanCandidates(DefaultApplicationContext.java:225)
	at io.micronaut.context.DefaultApplicationContext.transformIterables(DefaultApplicationContext.java:290)
	at io.micronaut.context.DefaultApplicationContext.findBeanCandidates(DefaultApplicationContext.java:226)
	at io.micronaut.context.DefaultBeanContext.isCandidatePresent(DefaultBeanContext.java:2972)
	at io.micronaut.context.DefaultBeanContext.containsBean(DefaultBeanContext.java:705)
	at io.micronaut.context.BeanDefinitionRegistry.containsBean(BeanDefinitionRegistry.java:355)
	at io.micronaut.context.RequiresCondition.matchesPresenceOfBeans(RequiresCondition.java:543)
	at io.micronaut.context.RequiresCondition.processPostStartRequirements(RequiresCondition.java:211)
	at io.micronaut.context.RequiresCondition.matches(RequiresCondition.java:107)
	at io.micronaut.context.AbstractBeanContextConditional.isEnabled(AbstractBeanContextConditional.java:50)
	at io.micronaut.context.AbstractBeanDefinition.isEnabled(AbstractBeanDefinition.java:86)
	at io.micronaut.context.DefaultBeanContext.lambda$findBeanCandidates$37(DefaultBeanContext.java:1701)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.micronaut.context.DefaultBeanContext.findBeanCandidates(DefaultBeanContext.java:1702)
	at io.micronaut.context.DefaultApplicationContext.findBeanCandidates(DefaultApplicationContext.java:225)
	at io.micronaut.context.DefaultBeanContext.lambda$findBeanCandidatesInternal$66(DefaultBeanContext.java:2782)
	at io.micronaut.core.util.clhm.ConcurrentLinkedHashMap.lambda$compute$0(ConcurrentLinkedHashMap.java:721)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
	at io.micronaut.core.util.clhm.ConcurrentLinkedHashMap.compute(ConcurrentLinkedHashMap.java:733)
	at io.micronaut.core.util.clhm.ConcurrentLinkedHashMap.computeIfAbsent(ConcurrentLinkedHashMap.java:710)
	at io.micronaut.context.DefaultBeanContext.findBeanCandidatesInternal(DefaultBeanContext.java:2782)
	at io.micronaut.context.DefaultBeanContext.getBeansOfTypeInternal(DefaultBeanContext.java:2820)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:1039)
	at io.micronaut.context.AbstractBeanDefinition.lambda$getBeansOfTypeForConstructorArgument$9(AbstractBeanDefinition.java:1165)
	at io.micronaut.context.AbstractBeanDefinition.resolveBeanWithGenericsFromConstructorArgument(AbstractBeanDefinition.java:1844)
	at io.micronaut.context.AbstractBeanDefinition.getBeansOfTypeForConstructorArgument(AbstractBeanDefinition.java:1160)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1009)
	at io.micronaut.liquibase.$$LiquibaseMigrationRunnerDefinition$InterceptedDefinition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1888)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2630)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2616)
	at io.micronaut.context.DefaultBeanContext.initializeEventListeners(DefaultBeanContext.java:1435)
	at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:2762)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:227)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:166)
	at io.micronaut.test.extensions.AbstractMicronautExtension.startApplicationContext(AbstractMicronautExtension.java:305)
	at io.micronaut.test.extensions.AbstractMicronautExtension.beforeClass(AbstractMicronautExtension.java:203)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.beforeAll(MicronautJunit5Extension.java:51)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$7(ClassBasedTestDescriptor.java:359)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:359)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:189)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:78)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:132)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:102)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:82)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:78)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy5.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:132)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:175)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:157)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Recursive update
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1760)
	at io.micronaut.core.util.clhm.ConcurrentLinkedHashMap.compute(ConcurrentLinkedHashMap.java:733)
	at io.micronaut.core.util.clhm.ConcurrentLinkedHashMap.computeIfAbsent(ConcurrentLinkedHashMap.java:710)
	at io.micronaut.context.DefaultBeanContext.findBeanCandidatesInternal(DefaultBeanContext.java:2782)
	at io.micronaut.context.DefaultBeanContext.getBeansOfTypeInternal(DefaultBeanContext.java:2820)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:1039)
	at io.micronaut.context.AbstractBeanDefinition.lambda$getBeansOfTypeForConstructorArgument$9(AbstractBeanDefinition.java:1165)
	at io.micronaut.context.AbstractBeanDefinition.resolveBeanWithGenericsFromConstructorArgument(AbstractBeanDefinition.java:1844)
	at io.micronaut.context.AbstractBeanDefinition.getBeansOfTypeForConstructorArgument(AbstractBeanDefinition.java:1160)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1012)
	at io.micronaut.runtime.context.$CompositeMessageSourceDefinition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1888)
	... 208 more
```

--- a/inject/src/main/java/io/micronaut/context/DefaultBeanContext.java
+++ b/inject/src/main/java/io/micronaut/context/DefaultBeanContext.java
@@ -117,7 +117,8 @@ public class DefaultBeanContext implements BeanContext {
     private final Map<CharSequence, Object> attributes = Collections.synchronizedMap(new HashMap<>(5));
 
     private final Map<BeanKey, Collection<Object>> initializedObjectsByType = new ConcurrentHashMap<>(50);
-    private final Map<BeanKey, Optional<BeanDefinition>> beanConcreteCandidateCache = new ConcurrentLinkedHashMap.Builder<BeanKey, Optional<BeanDefinition>>().maximumWeightedCapacity(30).build();
+    private final Map<BeanKey, Optional<BeanDefinition>> beanConcreteCandidateCache =
+            new ConcurrentLinkedHashMap.Builder<BeanKey, Optional<BeanDefinition>>().maximumWeightedCapacity(30).build();
     private final Map<Class, Collection<BeanDefinition>> beanCandidateCache = new ConcurrentLinkedHashMap.Builder<Class, Collection<BeanDefinition>>().maximumWeightedCapacity(30).build();
     private final Map<Class, Collection<BeanDefinitionReference>> beanIndex = new ConcurrentHashMap<>(12);
 
@@ -1089,30 +1090,36 @@ public class DefaultBeanContext implements BeanContext {
     }
 
     @Override
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings({"unchecked", "rawtypes"})
     public @NonNull
     <T> Optional<BeanDefinition<T>> findProxyTargetBeanDefinition(@NonNull Class<T> beanType, @Nullable Qualifier<T> qualifier) {
         ArgumentUtils.requireNonNull("beanType", beanType);
         Qualifier<T> proxyQualifier = qualifier != null ? Qualifiers.byQualifiers(qualifier, PROXY_TARGET_QUALIFIER) : PROXY_TARGET_QUALIFIER;
         BeanKey key = new BeanKey(beanType, proxyQualifier);
 
-        return (Optional) beanConcreteCandidateCache.computeIfAbsent(key, beanKey -> {
-            BeanRegistration<T> beanRegistration = singletonObjects.get(beanKey);
+        Optional beanDefinition = beanConcreteCandidateCache.get(key);
+        //noinspection OptionalAssignedToNull
+        if (beanDefinition == null) {
+            BeanRegistration<T> beanRegistration = singletonObjects.get(key);
             if (beanRegistration != null) {
                 if (LOG.isDebugEnabled()) {
                     LOG.debug("Resolved existing bean [{}] for type [{}] and qualifier [{}]", beanRegistration.bean, beanType, qualifier);
                 }
-                return Optional.of(beanRegistration.beanDefinition);
+                beanDefinition = Optional.of(beanRegistration.beanDefinition);
+            } else {
+                beanDefinition = findConcreteCandidateNoCache(
+                        null,
+                        (Class) beanType,
+                        proxyQualifier,
+                        true,
+                        false,
+                        false
+                );
             }
-            return findConcreteCandidateNoCache(
-                    null,
-                    (Class) beanType,
-                    proxyQualifier,
-                    true,
-                    false,
-                    false
-            );
-        });
+
+            beanConcreteCandidateCache.put(key, beanDefinition);
+        }
+        return beanDefinition;
     }
 
     @SuppressWarnings("unchecked")
@@ -1747,9 +1754,10 @@ public class DefaultBeanContext implements BeanContext {
             LOG.debug("Finding candidate beans for instance: {}", instance);
         }
         Collection<BeanDefinitionReference> beanDefinitionsClasses = this.beanDefinitionsClasses;
-        return beanCandidateCache.computeIfAbsent(instance.getClass(), aClass -> {
+        Class<?> beanType = instance.getClass();
+        Collection<BeanDefinition> beanDefinitions = beanCandidateCache.get(beanType);
+        if (beanDefinitions == null) {
             // first traverse component definition classes and load candidates
-
             if (!beanDefinitionsClasses.isEmpty()) {
 
                 List<BeanDefinition> candidates = beanDefinitionsClasses
@@ -1773,23 +1781,24 @@ public class DefaultBeanContext implements BeanContext {
                             .stream()
                             .filter(candidate ->
                                     !(candidate instanceof NoInjectionBeanDefinition) &&
-                                            candidate.getBeanType() == instance.getClass()
+                                            candidate.getBeanType() == beanType
                             )
                             .collect(Collectors.toList());
-                    return candidates;
+                    beanDefinitions = candidates;
                 }
                 if (LOG.isDebugEnabled()) {
                     LOG.debug("Resolved bean candidates {} for instance: {}", candidates, instance);
                 }
-                return candidates;
+                beanDefinitions = candidates;
             } else {
                 if (LOG.isDebugEnabled()) {
                     LOG.debug("No bean candidates found for instance: {}", instance);
                 }
-                return Collections.emptySet();
+                beanDefinitions = Collections.emptySet();
             }
-        });
-
+            beanCandidateCache.put(beanType, beanDefinitions);
+        }
+        return beanDefinitions;
     }
 
     /**
@@ -2453,16 +2462,28 @@ public class DefaultBeanContext implements BeanContext {
      * @param <T>             The bean generic type
      * @return The concrete bean definition candidate
      */
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings({"unchecked", "rawtypes"})
     private <T> Optional<BeanDefinition<T>> findConcreteCandidate(
             BeanResolutionContext resolutionContext,
             Class<T> beanType,
             Qualifier<T> qualifier,
             boolean throwNonUnique,
             boolean includeProvided) {
-        return (Optional) beanConcreteCandidateCache.computeIfAbsent(new BeanKey(beanType, qualifier), beanKey ->
-                (Optional) findConcreteCandidateNoCache(resolutionContext, beanType, qualifier, throwNonUnique, includeProvided, true)
-        );
+        BeanKey bk = new BeanKey(beanType, qualifier);
+        Optional beanDefinition = beanConcreteCandidateCache.get(bk);
+        //noinspection OptionalAssignedToNull
+        if (beanDefinition == null) {
+            beanDefinition = findConcreteCandidateNoCache(
+                    resolutionContext,
+                    beanType,
+                    qualifier,
+                    throwNonUnique,
+                    includeProvided,
+                    true
+            );
+            beanConcreteCandidateCache.put(bk, beanDefinition);
+        }
+        return beanDefinition;
     }
 
     private <T> Optional<BeanDefinition<T>> findConcreteCandidateNoCache(
@@ -2775,7 +2796,13 @@ public class DefaultBeanContext implements BeanContext {
 
     @SuppressWarnings("unchecked")
     private <T> Collection<BeanDefinition<T>> findBeanCandidatesInternal(BeanResolutionContext resolutionContext, Class<T> beanType) {
-        return (Collection) beanCandidateCache.computeIfAbsent(beanType, aClass -> (Collection) findBeanCandidates(resolutionContext, beanType, null, true));
+        @SuppressWarnings("rawtypes")
+        Collection beanDefinitions = beanCandidateCache.get(beanType);
+        if (beanDefinitions == null) {
+            beanDefinitions = findBeanCandidates(resolutionContext, beanType, null, true);
+            beanCandidateCache.put(beanType, beanDefinitions);
+        }
+        return beanDefinitions;
     }
 
     @SuppressWarnings("unchecked")
@@ -3310,7 +3337,7 @@ public class DefaultBeanContext implements BeanContext {
     /**
      * @param <T> The bean type
      */
-    private static class NoInjectionBeanDefinition<T> implements BeanDefinition<T>, BeanDefinitionReference<T> {
+    private final static class NoInjectionBeanDefinition<T> implements BeanDefinition<T>, BeanDefinitionReference<T> {
         private final Class<?> singletonClass;
         private final Map<Class<?>, List<Argument<?>>> typeArguments = new HashMap<>();
 

